### PR TITLE
Fix account details in-line error handling

### PR DESCRIPTION
**Pull request by daileytj**
Fixes blui-4284.

#### Changes proposed in this Pull Request:

- Fix account details in-line error handling
  - Only display error after initial onBlur
  - maintain error state when navigating between screens in the workflow

#### To Test:

- update verify code screen component line 50 and 57 to `if (name?.length > 2)` to see errors (note: after this update the error message will be wrong as we are by default returning an error message that says you need at least 1 character.)
- yarn start:example
- click create account & continue through the workflow
- when you reach the account details screen add invalid first name and last name & blur (error should display)
- nav back to the previous screen and forward again to the account details screen
- verify error displays

#### Any specific feedback you are looking for?

-

### CODE REVIEW

**Review by JeffGreiner-eaton — APPROVED** [review]

**Comment by joebochill** [comment]
What am I missing? Mine is not even preserving the initial values when I nav back.


https://github.com/etn-ccis/blui-react-workflows/assets/29152776/db2acfcf-44e1-47cb-82b2-280ae9ac0088

**Comment by daileytj** [comment]
https://github.com/etn-ccis/blui-react-workflows/assets/13989985/2e4f82e0-91df-498a-a34a-a74a330192c4

works on my machine ¯\_(ツ)_/¯

**Comment by joebochill** [comment]
Not sure how that's possible. I glanced at the code and we're never calling the 'update/onChange' callback functions for the field from the AccountDetailsBase. The screenData prop never gets set with account details info.

Your branch appears to be behind. Merge dev, blow away node modules & dist folder, and rerun the start command. See if it still works.



AccountDetailsScreenBase:
image

image

vs. CreateAccountScreenBase:
image

**Comment by JeffGreiner-eaton** [comment]
I did a fresh clone again and I don't notice the same issue mentioned above.

**Comment by joebochill** [comment]
Okay, I looked a little deeper and I know why it's not working.

First of all, how are you testing a custom validator? I'm testing it by passing in validator props to the screen on line 79 of AppRouter:
```
<AccountDetailsScreen firstNameValidator={(name): boolean | string => name.length > 5 ? true : 'Must be 5 chars'} lastNameValidator={(name): boolean | string => name.length > 10 ? true : 'Must be 10 chars'}/>
```

The issue is related to what I said before, but there's some nuance to it. 

The AccountDetails screen is not properly updating the screenData object.
- It's tracking a local state variable for first and last name, which is okay
- However, these variables are not being updated every time the base component input changes (we should be tapping into the base component onChange event and updating the full screen state variables on every keystroke)
- The full screen is ONLY updating its local state variable `onNext` and `onPrevious` (which is not correct — by this point there is no reason to update the local state because you're about to unmount and transition to another screen) and when the _default_ validator is executed (this is the main problem).
    - If you pass in a custom validator, it blows away the default, which was setting the only place the local state was changed. So the local state is NEVER updated with a custom validator.
    
Updating the state should not be tied in any way to the validator (in theory we should be able to support no validator at all to make these optional). The state should be updated when the underlying input changes and that's it. Whether it's valid or invalid, the local state should _always_ by in sync with the input.
    
image

**Comment by JeffGreiner-eaton** [comment]
> > I did a fresh clone again and I don't notice the same issue mentioned above.
> 
> ~Okay, I'll do another check on my end with a fresh clone to confirm...it could just be me 😵‍💫~
> 
> I did a brand new fresh clone into a new folder and it's exactly as seen in the recording from before. No state is maintained when going back to the password screen and then forward to account details.
> 
> FWIW, dev branch is doing exactly the same.

What else is modified in your local? I see the error messages are different than what Tom & I see by default...just wondering
image
image

**Comment by joebochill** [comment]
@JeffGreiner-eaton refer to my last comment.

Testing this by modifying the default validator function in the component is not a valid way to test this — we should be testing things the way users will use them (users will not be able to change the implementation of the component, they can only override the default through props).

**Comment by JeffGreiner-eaton** [comment]
@joebochill & @daileytj 
during the refactor here we noticed another bug and will write it up in the final fixes story/task. But if you pass in initial values for first and last name in App Router they currently do not work....

```                            <AccountDetailsScreen
                            initialFirstName='Mr.'
                            initialLastName='S'
                            firstNameLabel='your first name'
                            lastNameLabel='your last name'
                            firstNameValidator={(name): boolean | string => name.length > 5 ? true : 'Must be 5 chars'}
                            lastNameValidator={(name): boolean | string => name.length > 10 ? true : 'Must be 10 chars'} />```

File: login-workflow/src/new-architecture/screens/AccountDetailsScreen/AccountDetailsScreen.tsx [hunk 1]
```diff
@@ -20,7 +20,7 @@ export const AccountDetailsScreen: React.FC<AccountDetailsScreenProps> = (props)
             await actions().setAccountDetails({ firstName, lastName });
             void nextScreen({
                 screenId: 'AccountDetails',
-                values: { firstName, lastName },
+                values: { firstName: firstName, lastName: lastName },
```
Comment (daileytj) [inline login-workflow/src/new-architecture/screens/AccountDetailsScreen/AccountDetailsScreen.tsx:23]: You should be able to undo this change. No need to explicitly name them like this.

**Review by surajeaton — APPROVED** [review]
Looks good.
